example/common/const_body.hpp

```diff
@@ -14,29 +14,21 @@
 #include <beast/core/detail/type_traits.hpp>
 #include <boost/asio/buffer.hpp>
 #include <boost/optional.hpp>
-#include <boost/tti/has_type.hpp>
-#include <boost/tti/has_member_function.hpp>
-#include <boost/mpl/identity.hpp>
 #include <memory>
 #include <string>
 #include <utility>
 
-namespace const_detail
-{
-    BOOST_TTI_HAS_TYPE(value_type)
-    BOOST_TTI_HAS_TYPE(iterator)
-    BOOST_TTI_HAS_TYPE(size_type)
-    BOOST_TTI_HAS_TYPE(reference)
-    BOOST_TTI_HAS_MEMBER_FUNCTION(data)
-
-    template <typename T>
-    using is_container = boost::mpl::bool_<
-        has_type_value_type<T>::value &&
-        has_type_iterator<T>::value &&
-        has_type_size_type<T>::value &&
-        has_type_reference<T>::value &&
-        has_member_function_data<T, const typename T::value_type*, boost::mpl::vector<>, boost::function_types::const_qualified>::value
-    >;
+namespace detail {
+
+template<class T, class = void>
+struct is_const_container : std::false_type { };
+
+template< class T >
+struct is_const_container<T, beast::detail::void_t<
+    decltype( std::declval<T&>().size() ),
```

Comment:
Nice job on the traits! @pdimov showed me this trick for checking the return value:
```
    std::declval<std::size_t&>() = std::declval<T&>().size()
```

You can do something similar for checking the return value of `data`. 
```
    std::declval<std::void const*&>() = std::declval<T&>().data()
```